In [1]:
number =2 
print(type(number))

<class 'int'>


In [2]:
list= [1,2,4,5,6]
print(type(list))

<class 'list'>


In [3]:
class User:
    # The __init__ method is a special method called a constructor.
    def __init__(self, name, email, age, is_active): 
        # The parameters name, email, and age are values that must be provided when creating a new user object.
        #Counter (with capital C) = the class name (the blueprint)
        #counter1 (lowercase c) = a specific counter object made from that blueprint
        self.name = name
        self.email = email
        self.age = age
        self.is_active = is_active

    def display_profile(self):  # method shows how to access an object's attributes
        print(f"Name: {self.name}")
        print(f"Email: {self.email}")
        print(f"Age: {self.age}")
        print(f"Status: {'Active' if self.is_active else 'Inactive'}")

    def update_email(self, new_email):
        self.email = new_email
        print(f"Email updated to {new_email}")

    def deactivate_account(self):
        self.is_active = False
        print(f"Account for {self.name} has been deactivated")

    def reactivate_account(self):
        self.is_active = True
        print(f"Account for {self.name} has been reactivated")
#Creating objects from the User class
user1 = User("Alice Johnson", "alice@example.com", 28,True)
user2 = User("Bob Smith", "bob@example.com", 35, False )
user3 = User("Carol White", "carol@example.com", 42, True )


In [4]:
user1.display_profile()

Name: Alice Johnson
Email: alice@example.com
Age: 28
Status: Active


In [5]:
user2.display_profile()

Name: Bob Smith
Email: bob@example.com
Age: 35
Status: Inactive


In [6]:
user1.update_email('alice@gmail.com')



Email updated to alice@gmail.com


In [7]:
user2.deactivate_account()

Account for Bob Smith has been deactivated


In [8]:
user2.reactivate_account()

Account for Bob Smith has been reactivated


In [9]:
class Counter: # name of class 
    def __init__(self, start_value):
        # self refers to the new object being created
        # We're setting the 'count' attribute of this specific object
        self.count = start_value #An attribute is like a characteristic or piece of information that belongs to an object.
        print(f"Counter created with starting value: {start_value}")

    def increment(self):
        # self refers to whichever object called this method
        # We're modifying the 'count' attribute of that specific object
        self.count += 1      # self.count += value does: Takes the current value of self.count
                                #Adds value to it Assigns the result back to self.count like self.count = self.count+ value
        print(f"Count incremented to: {self.count}")

    def decrement(self):
        # Again, self refers to the specific object
        self.count -= 1
        print(f"Count decremented to: {self.count}")

    def get_count(self):
        # self lets us access the object's current count value
        return self.count

    def reset(self):
        # self lets us modify the object's attributes
        self.count = 0
        print("Counter reset to 0")

    def add_value(self, value):
        # self is used alongside other parameters
        # value is a parameter we explicitly pass
        # self is automatically passed by Python
        self.count += value
        print(f"Added{value}.Count is now:{self.count}:")

    
# Create two independent counter objects
counter1 = Counter(0)
counter2 = Counter(100)
# Output:
# Counter created with starting value: 0
# Counter created with starting value: 100

Counter created with starting value: 0
Counter created with starting value: 100


In [10]:
counter1.increment()
# Output: Count incremented to: 1

Count incremented to: 1


In [11]:
counter1.increment()
# Output: Count incremented to: 2
counter1.increment()
# Output: Count incremented to: 3

Count incremented to: 2
Count incremented to: 3


In [12]:
print(f"Counter 1 value: {counter1.get_count()}")
# Output: Counter 1 value: 3

# When we call counter2.increment(), Python passes counter2 as self
counter2.increment()
# Output: Count incremented to: 101

print(f"Counter 2 value: {counter2.get_count()}")
# Output: Counter 2 value: 101

# Demonstrating methods with additional parameters
counter1.add_value(10)
# Output: Added 10. Count is now: 13

counter2.add_value(50)
# Output: Added 50. Count is now: 151

# Resetting one counter doesn't affect the other
counter1.reset()
# Output: Counter reset to 0

print(f"Counter 1: {counter1.get_count()}")  # Output: Counter 1: 0
print(f"Counter 2: {counter2.get_count()}")  # Output: Counter 2: 151


Counter 1 value: 3
Count incremented to: 101
Counter 2 value: 101
Added10.Count is now:13:
Added50.Count is now:151:
Counter reset to 0
Counter 1: 0
Counter 2: 151


In [13]:
#class is a blueprint
# object : An instance of a class
# creating a class
class Student:
    def __init__(self, name, major):
        #Instance attributes
        self.name = name
        self.major = major
        self.grades = []
    # Instance method 
    def add_grade(self, grade):
        self.grades.append(grade) # append means add at the end

    #Instance method

    def get_average(self):
           # What condition do I want to check? Example: “Is the list empty?”
        if not self.grades:
            #What should happen if the condition is true? Example: “Return 0
            return 0
        return sum(self.grades)/ len(self.grades)
    # Instance method
    def get_info(self):
        avg = self.get_average()
        return f"{self.name}, (Major: {self.major}), Average: {avg: .2f}"




    
        


In [14]:
alice = Student("ALice", "Data Science")
alice.add_grade(85)
alice.add_grade(92)
alice.add_grade(78)

alice.get_info()
print("="* 50)
alice.grades # Direct access to attribut



[85, 92, 78]

In [15]:
# Create Encapsulation attribute with double underscore
class BankAccount:
    def __init__(self, account_number, balance):
        self.account_number = account_number  # Public
        self.__balance = balance              # Private


account = BankAccount("ACC123", 1000)
print(account.account_number)  # Output: ACC123

# Trying to access private attribute directly
# print(account.__balance)  # This will cause an AttributeError


    


        

ACC123


# Controlling Access with Methods
* Getter methods (also called accessor methods) allow controlled read access to private attributes.
  
* They return the value of a private attribute without exposing the attribute itself for direct manipulation.

In [16]:
class BankAccount:
    def __init__(self, account_number, balance):
        self.account_number = account_number
        self.__balance = balance
    def get_balance(self):
        return self.__balance
    
        


In [17]:
#account = Variable name (the label you choose)
# = = Assignment operator (puts the object into the variable)
# BankAccount("ACC123", 1000) = Object creation (makes a new BankAccount)
account= BankAccount("ACC123", 1000)
#Method Call Syntax: account.get_balance()
# account = The object (instance) you're working with
# . = The dot operator (connects object to method)
#get_balance = The method name
# () = Parentheses (required to actually call the method)
current_balance = account.get_balance()
print(f"Balance: ${current_balance}") 

Balance: $1000


## Setter methods (also called mutator methods) allow controlled write access to private attributes.
* They let you modify a private attribute's value.
* But you can include validation logic to ensure only valid values are set.

In [18]:
class BankAccount:
    def __init__(self, account_number, balance):
        self.account_number = account_number 
        self.__balance = balance
    def get_balance(self):
        return self.__balance
    def set_balance(self, new_balance):
        if new_balance>0:
            self.__balance = new_balance
            print(f"Balance updated to ${self.__balance}") 
        else:
            print( "Error: Balance cannot be negative")
        

        
 
    

In [19]:
account = BankAccount("ACC123", 1000)
account.set_balance(1500)  # Output: Balance updated to $1500
account.set_balance(-500)  # Output: Error: Balance cannot be negative

print(account.get_balance())  # Output: 1500


Balance updated to $1500
Error: Balance cannot be negative
1500


# Why Use Getters and Setters?
* Validation and Data Integrity: Setter methods let you validate data before it's stored. You can check that values are in acceptable ranges, that strings aren't empty, that numbers are positive when they should be, or any other business rule that applies.

In [20]:
class Product:
    def __init__(self, name, price):
        self.__name = name
        self.__price = price
    def get_price(self):
        return self.__price
    def set_price(self, new_price):
        if new_price>0:
            self.__price = new_price
        else:
            print("Error: Price must be positive")
    def get_name(self):
        return self.__name
    def set_name(self, new_name):
        if len(new_name.strip()) >0:
            self.__name = new_name
        else:
            print("Erorr: Name cannot be empty")
        
    
    

* Maintaining Consistency: Sometimes changing one attribute requires updating others to keep the object in a valid state. Methods let you coordinate these related changes.

In [21]:
class Reclangle:
    def __init__( self, width, height):
        self.__width = width
        self.__height = height
        self.__area = width * height
    def set_width(self, new_width):
       if new_width >0:
           self.__width = new_width
           self.__area = self.__width * self.__height
           
    def set_width(self, new_height):
       if new_height >0:
           self.__height= new_height
           self.__area = self.__width * self.__height
    def get_area(self):
        return self.__area
        
    
      
        

* Flexibility to Change Implementation: If you use methods to access data, you can change how data is stored internally without breaking code that uses your class. The external interface stays the same even if the internal details change.
* Code that calls get_full_name() works with both versions, even though the internal storage changed completely.


In [22]:
# Version 1: Store full name as one attribute
class Person:
    def __init__(self, full_name): 
        self.__full_name = full_name
    def get_full_name(self):
        return self.__full_name
# Version 2: Store first and last name separately
# But keep the same interface    
class Person:
     def __init__ (self, first_name, last_name): 
         self.__first_name = first_name
         self.__last_name = last_name 

     def get_full_name(self):
         return f"{self.__first_name} {self.__last_name}"


* Adding Computed Properties: Sometimes what looks like an attribute is actually calculated on demand. Methods let you compute values when needed without storing them.

In [23]:
class Circle:
    def __init__(self, radius):
        self.__radius = radius
    
    def get_radius(self):
        # area is computed udsing the formula A = πr²
        return 3.14159*self.__radius**2
    def get_circumfrence(self): 
        # circumfrence is computed using the formula C = 2πr
        return 2*3.14159*self.__radius



        

# Buliding a complete Example 
###  Building a more realistic class that demonstrates encapsulation principles. We'll create a Student class that manages student information with appropriate protection.

In [24]:
class School:
    def __init__(self, student_id, name, age):
        self.__student_id = student_id # private attribute
        self.__name = name
        self.__age = age 
        self.__grades = [] # private attribute
    def get_student_id(self):
        return self.__student_id 
    # When we want an attribute to be truly private and read-only, 
    # we don't provide a setter method. 
    # This is a key principle of encapsulation.
    def get_name(self):
        return self.__name
    def set_name(self, new_name):
        if len(new_name.strip()) > 0:
            # The strip() method removes whitespace characters (spaces, tabs, newlines) 
            # from the beginning and end of a string.
            # Without strip(), someone could set a name to just spaces,
            #  which would pass a simple length check but isn't really a valid name:
            self.__name = new_name
            return True
        else:
            return False 
    def get_age(self):
        return self.__age
    def set_age(self, new_age):
        if 0 < new_age < 120:
            self.__age = new_age
            return True
        else:
            return False
    def add_grade(self, grade):
        if 0 <= grade <= 100:
            self.__grades.append(grade)
            print(f"Grade {grade} added.")
            return True
        else:
            print("Error: Grade must be between 0 and 100.")
            return False
    def get_grades(self):
        return self.__grades.copy()
         # Return a copy of copy the grades list to prevent external modification
         # The .copy() method creates a new list with the same elements as the original,
         #  but it's a completely separate list in memory.
    def calculate_average(self): 
        if len(self.__grades) == 0:
            return 0
        else:
            return sum(self.__grades) / len(self.__grades)
 #Finally, add a method to display student information:
    def display_info(self):
        avg = self.calculate_average()
        print(f"Student ID: {self.__student_id}")
        print(f"Name: {self.__name}")
        print(f"Age: {self.__age}")
        print(f"Number of grades: {len(self.__grades)}")
        print(f"Average Grade: {avg:.2f}")

        

##### __why self.grades = [] appears in the __init__ method even though "grades" isn't a parameter. The key insight is that not every attribute needs to come from a parameter.__ 
__Here's why:__

Parameters vs. Initial Values
When we create a Student object, we need to provide some basic information (student_id, name, age), but a new student starts with no grades yet. So we initialize __grades as an empty list.

Think of it like this: When a student enrolls in school, they have:

A student ID (provided)
A name (provided)
An age (provided)
An empty grade book (starts empty, gets filled later)
How Do We Know to Add This?
You know to add self.__grades = [] because you're thinking ahead about what the Student class will need to do. Ask yourself:

What data will this object need to store over its lifetime?
What will start with initial values vs. what gets provided during creation?

In [25]:
# Create an Education object
school = School("S12345", "Alice Johnson", 20)

# Access public interface
print(school.get_name())  # Output: Alice Johnson

# Add grades through controlled method
school.add_grade(85)  # Output: Grade 85 added
school.add_grade(92)  # Output: Grade 92 added
school.add_grade(78)  # Output: Grade 78 added

# Try to add invalid grade
school.add_grade(105)  # Output: Error: Grade must be between 0 and 100

# View information
school.display_info()
# Output:
# Student ID: S12345
# Name: Alice Johnson
# Age: 20
# Number of grades: 3
# Average: 85.00


Alice Johnson
Grade 85 added.
Grade 92 added.
Grade 78 added.
Error: Grade must be between 0 and 100.
Student ID: S12345
Name: Alice Johnson
Age: 20
Number of grades: 3
Average Grade: 85.00


# __Inheritance and Code Reusability__
*  It allows you to create a new class based on an existing class, inheriting all the attributes and methods from the original class while adding new functionality or modifying existing behavior.
* Inheritance  lets you define shared attributes and methods once in a parent class, then create child classes that automatically receive that functionality.
* The child classes can then add their own unique features without duplicating the common code.
* This approach keeps your code organized, reduces duplication, and makes maintenance much easier.
* The parent class (also called the base class or superclass) contains the general, shared functionality. 
* The child classes (also called derived classes or subclasses) inherit everything from the parent and then add or customize as   needed. 
* This creates a hierarchical relationship where specialized classes build upon more general ones.



In [26]:
# create a  parent clas
class Person:
    def __init__ (self, name, age, email):
        self.name = name
        self.age = age
        self.email =email

    def display_info(self):
        print(f"Name: {self.name}")
        print(f"Age: {self.age}")
        print(f"Email: {self.email}")
    def update_email(self, new_email):
        self.email = new_email
        print("Email updated successfully.")




# __Creating a Child Class__
* To create a child class that inherits from a parent class, we specify the parent class name in parentheses after the child class name. 
* This tells Python that the new class should inherit all attributes and methods from the parent.


In [27]:
# creating a child class
class Students(Person):
    def __init__(self, name, age,email,student_number):
        super().__init__(name, age, email ) # Call the parent class's constructor to initialize inherited attributes
        self.student_number = student_number
        self.course = []
    
    def enroll_course(self, new_course):
        self.course.append(new_course)  # Fixed: now matches the attribute name
        print(f"{self.name} enrolled in {new_course} successfully.")
   
    





## Let's examine what's happening here.
 * The line class Student(Person): declares that Student is a child class of Person. This means Student automatically inherits all attributes and methods from Person.

* The super().__init__(name, age, email) line is crucial. The super() function gives you access to the parent class. Here, we're calling the parent's __init__ method to handle setting up the name, age, and email attributes. This avoids duplicating that initialization code in the child class.

* After calling the parent's constructor, we add attributes that are specific to students: student_id and courses. These additional attributes exist only in Student objects, not in general Person objects.

*  Notice that student has access to both the display_info() and update_email() methods from Person, plus the enroll() method that's specific to Student. The child class has everything the parent has, plus its own additions.

In [28]:
# Test the inheritance
students = Students("John Doe", 20, "john@university.edu", "STU001")
students.enroll_course("Python Programming")
students.update_email("alice.j@school.edu")
students.display_info()

John Doe enrolled in Python Programming successfully.
Email updated successfully.
Name: John Doe
Age: 20
Email: alice.j@school.edu


## __Multiple Child Classes__


In [29]:
class Teacher(Person):
    def __init__(self, name, age, email, employee_id, subject):
        super().__init__(name, age, email)
        self.employee_id = employee_id
        self.subject = subject
        self.classes = [] # an instance attribute that stores a list of class names for a particular teacher object. 
    def assign_class(self, new_class):
        # Append the new class to the classes lisThe .
        # append() method adds a new item to the end of a list.
        # It's like adding a new course to the bottom of your class schedule.

        self.classes.append(new_class)
        #Breaking down self.classes.append(class_name):
        # self: Refers to the current student object
        # .classes: The list attribute that stores all enrolled classes
        # .append(): The method that adds to the end of the list
        # class_name: The new class being added (like "Math 101" or "History")
        print(f"{self.name} assigned to teach {new_class} successfully.")

class Administrator(Person):
    def __init__(self, name, age, email, admin_id, department):
        super().__init__(name, age, email)
        self.admin_id = admin_id
        self.department = department
    def make_announcement(self, message): 
        print(f"Announcement from {self.name} ({self.department}): {message}")




### All three above classes share the basic __Person__ functionality, but each adds its own specific features. If you need to modify how display_info() works, you change it once in Person, and all three child classes automatically get the update.

In [30]:
# Using all three classes - Test the inheritance
students = Students("Alice", 20, "alice@school.edu", "S001")
teacher = Teacher("Bob", 35, "bob@school.edu", "T001", "Math")
admin = Administrator("Carol", 45, "carol@school.edu", "A001", "Academics")

# All three can use Person methods
students.display_info()
print("="*30)
teacher.display_info()
print("="*30)
admin.display_info()
print("="*30)

# Each has unique methods
students.enroll_course("Calculus")
teacher.assign_class("Calculus 101")
admin.make_announcement("Classes start next Monday")

Name: Alice
Age: 20
Email: alice@school.edu
Name: Bob
Age: 35
Email: bob@school.edu
Name: Carol
Age: 45
Email: carol@school.edu
Alice enrolled in Calculus successfully.
Bob assigned to teach Calculus 101 successfully.
Announcement from Carol (Academics): Classes start next Monday


###__The super() Function__
* The most common use is in constructors, where you call super().__init__() to initialize the parent's attributes before adding child-specific attributes:





In [31]:
class Employee(Person):
    def __init__(self, name, age, email, employee_id, salary):
        super().__init__(name, age, email)  # Initialize parent attributes
        self.employee_id = employee_id
        self.salary = salary  # Add child-specific attribute

#  We can also use super() to call other parent methods from within child methods
class Manager(Employee):
    def __init__(self, name, age, email, employee_id, salary, department):
        super().__init__(name, age, email, employee_id, salary)  # Call Employee's constructor
        self.department = department
        self.team_members =[] # Initialize an empty list for team members
    def display_info(self):
        super().display_info() # call parents display_info method
        print(f"Department: {self.department}")
        # Display team members with conditional formatting
        # If team_members list has items: join them with commas
        # If team_members list is empty: display 'None'
        # ', '.join(self.team_members) - takes list items and joins them with ", " separator
        # The if/else is a ternary operator: value_if_true if condition else value_if_false
        print(f"Team Members: {', '.join(self.team_members) if self.team_members else 'None'}")
        print(f"Team size: {len(self.team_members)}")

## Why We Don't Use `.append()` Here

The line `', '.join(self.team_members)` is **displaying** existing data, not **adding** new data to the list.

### Key Differences:

**`.append()`** - Used for **ADDING** items TO a list:
- `self.team_members.append("John")` → Adds "John" to the team_members list
- **Purpose:** Modify the list by adding new items
- **Result:** The list grows by one item

**`.join()`** - Used for **DISPLAYING** items FROM a list:
- `', '.join(self.team_members)` → Takes existing items and creates a readable string
- **Purpose:** Convert list contents into a formatted string for display
- **Result:** Returns a string, doesn't change the original list

**Think of it this way:** `.append()` puts things IN the box, `.join()` reads what's already IN the box and makes it pretty to display.

## What Does `len()` Do?

The `len()` function **counts** the number of items in a list (or other collections).

### Syntax:
```python
len(list_name)  # Returns the number of items in the list
```

### Examples with team_members:

```python
# Empty list
team_members = []
len(team_members)  # Returns: 0

# List with 2 members
team_members = ["Alice", "Bob"]
len(team_members)  # Returns: 2

# List with 5 members
team_members = ["Alice", "Bob", "Carol", "David", "Eve"]
len(team_members)  # Returns: 5
```

### In the context of `print(f"Team size: {len(self.team_members)}")`:

- **`self.team_members`** - The list containing team member names
- **`len(self.team_members)`** - Counts how many names are in the list
- **`f"Team size: {len(self.team_members)}"`** - Creates a formatted string showing the count

**Think of it this way:** `len()` is like counting how many people are in a room - it gives you the total number, not their names.

In [32]:
# Real Example:

team_members = ["Alice", "Bob", "Carol"]
print(f"Team size: {len(team_members)}")

team_members = []
print(f"Team size: {len(team_members)}")


Team size: 3
Team size: 0


In [33]:
class Person:
    def __init__(self, name, age): # constructor method
        self.name = name # instance attribute
        self.age = age # instance attribute

    def introduce(self): # instance method
        print(f"Hi, I'm {self.name}") # method body

class Student(Person):# child class inheriting from Person
    def __init__(self, name, age, major): # constructor method
        super().__init__(name, age) # call parent constructor
         # initialize inherited attributes  
        self.major = major # child-specific attribute

    def introduce(self): # overriding method
        
        print(f"Hi, I'm {self.name}, a {self.major} student") # method body

## Benefits of Method Overriding

**Method overriding** allows child classes to provide their own specific implementation of a method that already exists in the parent class.

### Key Benefits:

#### 1. **Customized Behavior for Different Types**
- Each child class can behave differently while maintaining the same interface
- Example: Both `Person` and `Student` have `introduce()`, but each works appropriately for its type

```python
person = Person("Alice", 25)
student = Student("Bob", 20, "Computer Science")

person.introduce()    # Output: "Hi, I'm Alice"
student.introduce()   # Output: "Hi, I'm Bob, a Computer Science student"
```

#### 2. **Polymorphism** 
- You can treat different objects the same way, but they behave according to their specific type
- Same method call, different behavior based on the actual object type

```python
people = [Person("Alice", 25), Student("Bob", 20, "CS"), Student("Carol", 19, "Math")]

for person in people:
    person.introduce()  # Each will introduce themselves appropriately
```

#### 3. **Maintain Common Interface**
- All objects respond to the same method calls
- Makes code more flexible and easier to maintain
- You can add new types without changing existing code

#### 4. **Extend Rather Than Replace**
- Child classes can build upon parent functionality
- Use `super()` to call the parent method first, then add extra behavior

```python
def introduce(self):
    super().introduce()  # Call parent's introduce first
    print(f"I'm studying {self.major}")  # Add student-specific info
```

**Think of it this way:** Like having the same remote control that works differently for TV, stereo, and DVD player - same buttons, but each device responds in its own way.

In [34]:
# USage
person = Person("John", 30)
person.introduce()  # Output: Hi, I'm John

student = Student("Alice", 20, "Computer Science")
student.introduce()  # Output: Hi, I'm Alice, a Computer Science

Hi, I'm John
Hi, I'm Alice, a Computer Science student


In [35]:
class Teacher(Person):
    def __init__(self, name, age, subject):
        super().__init__(name, age)
        self.subject = subject

    def introduce(self):
        # We can also partially override by calling the parent method and then adding more:
        super().introduce()  # Call parent version first
        print(f"I teach {self.subject}")

In [36]:
teacher = Teacher("Mr. Smith", 40, "Mathematics")
teacher.introduce()

Hi, I'm Mr. Smith
I teach Mathematics


#**When to Override Methods** 
* You should override methods when the child class needs different behavior than the parent provides. 
**Common scenarios include:**

* Specialization: The child class needs a more specific version of a general parent method. A Dog class might override an Animal class's make_sound() method to return "Bark!" instead of a generic sound.

* Additional Functionality: The child needs to do everything the parent does plus more. Use super() to call the parent method, then add the extra functionality.

* Completely Different Behavior: Sometimes the child class needs entirely different logic. In these cases, you override without calling super().


In [37]:
# Here's an example showing different override approaches:
class BankAccount:
    def __init__(self, account_number, balance):
        self.account_number = account_number
        self.balance = balance
    def calculate_interst(self):
        return self.balance * 0.01 # 1% interest
class SavingAccount(BankAccount):
    def calculate_interest(self):
        return self.balance * 0.03 # 3% interest
class CheckingAccount(BankAccount):
    def calculate_interst(self):
        return 0 # No interest for checking accounts
 


In [38]:
#usage
bank_account = BankAccount("ACC001", 5000)
savings = SavingAccount("SAV123", 1000)
checking = CheckingAccount("CHK456", 2000)  
print(f"Checking Account Interest: ${checking.calculate_interst():.2f}")
print(f"Saving Account Interest: ${savings.calculate_interest():.2f}")
print(f"Account Interest: ${bank_account.calculate_interst():.2f}") 

Checking Account Interest: $0.00
Saving Account Interest: $30.00
Account Interest: $50.00


***Inheritance can extend beyond two levels.***
* A child class can itself become a parent to another class, creating an inheritance hierarchy. However, you should generally keep hierarchies shallow—two or three levels at most — to avoid complexity.

In [39]:
class Vehicle: # parent class
    def __init__(self, brand, model):# constructor method
        self.brand = brand # instance attribute
        self.model = model # instance attribute

    def start(self):# instance method 
        print(f"{self.brand} {self.model} is starting") #  method body 

class Car(Vehicle):# child class inheriting from Vehicle
    def __init__(self, brand, model, num_doors):# constructor method
        super().__init__(brand, model) # call parent constructor 
        self.num_doors = num_doors # child-specific attribute

class ElectricCar(Car):# child class inheriting from Car
    def __init__(self, brand, model, num_doors, battery_capacity):# constructor method
        super().__init__(brand, model, num_doors)# call parent constructor
        self.battery_capacity = battery_capacity # child-specific attribute

    def charge(self): # instance method
        print(f"Charging {self.brand} {self.model}") # method body

In [40]:
# The ElectricCar has access to attributes and methods from both Car and Vehicle through the inheritance chain.
tesla = ElectricCar("Tesla", "Model 3", 4, "75 kWh")
tesla.start()   # Inherited from Vehicle
tesla.charge()  # Specific to ElectricCar
print(tesla.num_doors)  # Inherited from Car

Tesla Model 3 is starting
Charging Tesla Model 3
4


## How `self` Works Across Inheritance - Same Object, Shared Attributes

**Important Concept:** When you create a child class object, `self` refers to the **SAME SINGLE OBJECT** throughout the entire inheritance chain. All attributes belong to this one object.

### Example Breakdown:

```python
tesla = ElectricCar("Tesla", "Model 3", 4, "75 kWh")
```

**What happens during object creation:**

1. **ElectricCar.__init__()** runs first
   - `self.battery_capacity = "75 kWh"` (added to the tesla object)
   - Calls `super().__init__(brand, model, num_doors)`

2. **Car.__init__()** runs next
   - `self.num_doors = 4` (added to the SAME tesla object)
   - Calls `super().__init__(brand, model)`

3. **Vehicle.__init__()** runs last  
   - `self.brand = "Tesla"` (added to the SAME tesla object)
   - `self.model = "Model 3"` (added to the SAME tesla object)

### Result: ONE Object with ALL Attributes

```python
# The tesla object now contains:
tesla.brand = "Tesla"           # From Vehicle
tesla.model = "Model 3"         # From Vehicle  
tesla.num_doors = 4             # From Car
tesla.battery_capacity = "75 kWh"  # From ElectricCar
```

**Key Point:** There's only **ONE** `tesla` object. The `self` parameter in all three `__init__` methods refers to this same object. Each class adds its attributes to the same object instance.

**Think of it like building a house:** Each contractor (parent class) adds their part to the SAME house, not separate houses.

**Checking Inheritance Relationships**
* Python provides built-in functions to check inheritance relationships between classes and objects.

* The isinstance() function checks if an object is an instance of a specific class or any of its parent classes:

In [41]:
students = Students("Alice", 20, "alice@school.edu", "S001")

print(isinstance(students, Students))  # Output: True
print(isinstance(students, Person))   # Output: True
print(isinstance(students, Teacher))  # Output: False

True
False
False


**The issubclass() function checks if one class is a child of another:**

In [42]:
print(issubclass(Students, Person))   # Output: True
print(issubclass(Teacher, Person))   # Output: True
print(issubclass(Students, Teacher))  # Output: False

False
True
False


**When to Use Inheritance**
* Inheritance is powerful but should be used judiciously. Use inheritance when there's a clear "is-a" relationship. A student IS A person. A car IS A vehicle. These are natural inheritance relationships.

**Good uses of inheritance:**

* Creating specialized versions of a general class (Teacher is a Person with teaching-specific features)
* Sharing common functionality across related classes (Dog, Cat, and Bird all inherit from Animal)
* Building frameworks where users extend base classes with custom behavior

**When not to use inheritance:**

* When composition makes more sense (a Car HAS AN Engine, but shouldn't inherit from Engine)

* When classes are unrelated despite some shared behavior
* When a simple function or module would suffice
Consider this example where inheritance isn't appropriate:



In [43]:
# Consider this example where inheritance isn't appropriate:

# Poor design - FileReader shouldn't inherit from File
class File:
    def __init__(self, filename):
        self.filename = filename

class FileReader(File):  # Bad: Reader isn't a type of File
    def read(self):
        pass
# Better design using composition:
# Better design - FileReader uses a File
class File:
    def __init__(self, filename):
        self.filename = filename

class FileReader:
    def __init__(self, file):
        self.file = file  # Composition: Reader HAS A File

    def read(self):
        pass

**Property Decorators and Pythonic Encapsulation**

* Decorators are particularly useful in OOP because they let you add functionality to your methods without changing how users of your class interact with those methods. 

**what decorators are**
* A decorator is a special syntax in Python that modifies the behavior of functions or methods. 

* You recognize decorators by the @ symbol followed by the decorator name, placed on the line directly above a function or method definition.

**Type of decorator**

 1- **@property decorator**
* allows us to create attributes that look and behave like regular attributes but actually run methods behind the scenes.
* specifically transforms methods into attribute-like access, making your code cleaner and more intuitive.
* converts a method into a "getter" that can be accessed like an attribute. Instead of calling a method with parentheses, you access it as if it were a simple attribute.

In [44]:
# Here's a comparison showing the difference:
# Tradintional getter method
class OldStyle:
     def __init__(self, value): # Initialize with a value
         self.__value = value  # private attribute

     def get_value(self):
         return self.__value
obj = OldStyle(10)
print(obj.get_value())  # Must use parentheses

# @Property decorator method
class NewStyle:
    def __init__(self, value): # Initialize with a value
        self.__value = value # private attribute
    @property
    def value(self):
        return self.__value
obj2 = NewStyle(20)
print(obj2.value)  # no parentheses needed
    



10
20


In [45]:
# complex example
class Temperature:
    def __init__(self, celsius):
        self.__celsius = celsius
    @property
    def celsius(self):
        return self.__celsius
    @property
    def farenheit(self):
        return (self.__celsius * 9/5) + 32
    @property
    def kelvin(self):
        return self.__celsius +273.15

temp = Temperature(25)
print(temp.celsius)     # 25            
print(temp.farenheit)   # 77.0
print(temp.kelvin)      # 298.15


25
77.0
298.15


**Note**
* The fahrenheit and kelvin properties are computed on-the-fly whenever you access them. They don't exist as stored attributes — they're calculated from the celsius value each time. This is more efficient than storing multiple representations of the same data.


## Why Protected Attributes Use Single Underscore `_`

The **single underscore prefix** (`_attribute`) is Python's naming convention to indicate **protected attributes**.

**Why Use the Underscore?**

1. **Clear Visual Distinction**
   - Easy to spot which attributes are internal vs public
   - Instantly tells you "be careful with this"

2. **Python Community Standard**
   - All Python developers understand this convention
   - Used in Python standard library and major frameworks

3. **IDE and Tool Support**
   - Many editors highlight `_attributes` differently
   - Static analysis tools can warn about accessing protected members

**Remember:** The underscore is a **communication tool** - it tells other programmers "this is internal, use the public interface instead."

In [46]:

class Example:
    def __init__(self):
        self.public_attr = "Anyone can use"      # No underscore = Public
        self._protected_attr = "Internal use"    # Single _ = Protected
        self.__private_attr = "Very restricted"  # Double __ = Private
class BankAccount:
    def __init__(self, balance):
        self.account_number = "12345"    # Public - customers can see
        self._balance = balance          # Protected - only for internal bank systems
        self.__pin = "1234"             # Private - extremely sensitive

# Usage
account = BankAccount(1000)
print(account.account_number)  # ✅ Fine - it's public
print(account._balance)        # ⚠️ Works but signals "internal use only"

12345
1000


In [47]:
# Using this class:
account = BankAccount(1000)

# Getting the balance (uses @property)
print(account.balance)  # Output: 1000

# Setting the balance (uses @balance.setter)
account.balance = 1500
print(account.balance)  # Output: 1500

# Invalid value triggers validation
account.balance = -50 # Raises ValueError


AttributeError: 'BankAccount' object has no attribute 'balance'

In [ ]:
# Building a Complete Example
class Person1:
    def __init__ (self, first_name, last_name, age):
        self.__first_name = first_name
        self.__last_name = last_name 
        self.__age = age
    @property
    def first_name(self):
        return self.__first_name
    @first_name.setter # setter method
    def first_name(self, new_first_name): # setter method
        if len(new_first_name.strip()) > 0: # validation check
        
            self.__first_name = new_first_name # set the private attribute
        else:
            raise ValueError("First name cannot be empty") # raise an error
    @property # last_name getter method
    def last_name(self): # getter method
        return self.__last_name # return the private attribute

    @last_name.setter # last_name setter method
    def last_name(self, name): # setter method
        if len(name.strip()) > 0: # validation check
            self.__last_name = name.strip() # set the private attribute
        else:
            raise ValueError("Last name cannot be empty")

    @property
    def age(self): 
        return self.__age

    @age.setter
    def age(self, value):
        if 0 <= value <= 120:
            self.__age = value
        else:
            raise ValueError("Age must be between 0 and 120")
    
    @property
    def full_name(self):
        return f"{self.__first_name} {self.__last_name}"

    @full_name.setter
    def full_name(self, name):
        parts = name.split()
        if len(parts) >= 2:
            self.__first_name = parts[0]
            self.__last_name = " ".join(parts[1:])
        else:
            raise ValueError("Full name must include first and last name")

In [ ]:
person = Person1("John", "Doe", 30)

# Read properties
print(person.first_name)  # Output: John
print(person.full_name)   # Output: John Doe

# Modify properties
person.age = 31
person.last_name = "Smith"
print(person.full_name)   # Output: John Smith

# Set full name (splits into first and last)
person.full_name = "Jane Anderson"
print(person.first_name)  # Output: Jane
print(person.last_name)   # Output: Anderson

# Validation works automatically
person.age = 150  # Raises ValueError
person.first_name = ""  # Raises ValueError

John
John Doe
John Smith
Jane
Anderson


ValueError: Age must be between 0 and 120

In [ ]:
class Product1:
    def __init__(self, name, price, quantity):
        self.__name = name
        self.__price = price
        self.__quantity = quantity
        self.__product1_id = id(self)  # Unique ID

    @property
    def product1_id(self):
        return self.__product1_id

    # No setter - product_id is read-only

    @property
    def total_value(self):
        return self.__price * self.__quantity

    # No setter - total_value is computed, not stored

In [ ]:
# Usage
product = Product1("Laptop", 999.99, 10)

print(product.product1_id)    # Can read
print(product.total_value)   # Can read

# These will raise AttributeError
product.product1_id = 12345  # Error: can't set attribute
product.total_value = 5000  # Error: can't set attribute

1457784658240
9999.9


AttributeError: property 'product1_id' of 'Product1' object has no setter

**The id() function** returns where Python stored your object in the computer's memory. This number is:

* Unique for each object
* Different every time you run the program
* Large because it's a memory address
* Not user-friendly but guaranteed to be unique

The large number 1457784656224 is the memory address of your Product1 object.

What's happening:

self.__product1_id = id(self) gets the object's memory location
id() returns where Python stored the object in computer memory
This creates a unique but unfriendly identifier
Your output breakdown:

1457784656224 = Memory address (unique object identifier)
9999.9 = Total value calculation (999.99 × 10 = 9999.90)

**Advantages of Property Decorators**
 * Property decorators provide several significant advantages over traditional getter and setter methods.

    * **Cleaner Syntax**: Your code reads more naturally. Compare account.balance versus account.get_balance(). The property version looks cleaner and more intuitive.
    
    * **Backward Compatibility**: You can start with a simple public attribute and later convert it to a property with validation without breaking existing code:

    * **Computed Attributes**: Properties let you calculate values on demand rather than storing them, which saves memory and ensures values are always current:

    * **Lazy Initialization**: Properties can defer expensive computations until the value is actually needed:


In [ ]:
# Version 1: Simple public attribute
class User1:
    def __init__(self, email):
        self.email = email

# Version 2: Add validation without breaking code
class User1:
    def __init__(self, email):
        self.__email = email

    @property
    def email(self):
        return self.__email

    @email.setter
    def email(self, value):
        if "@" in value:
            self.__email = value
        else:
            raise ValueError("Invalid email")

#Code that uses user.email works with both versions,
#  but Version 2 adds validation protection.

In [ ]:
# Lazy Initialization: What is this class doing? 
# The DataAnalyzer class analyzes data,but it's designed to be efficient 
# by only doing the expensive analysis work when it's actually needed.


class DataAnalyzer: 
    def __init__(self, data): # The __init__ method
        self.__data = data # Stores the data to be analyzed
        self.__analysis = None # Starts as None (no analysis done yet)

# @property: This makes analysis behave like an attribute, 
# but it's actually a method. So you can write analyzer. analysis 
# instead of analyzer.analysis().
    @property
    def analysis(self):
       # This is the smart part:First time: If no analysis has been done (None),
       #  do it now and save the resultLater times:
       #Just return the saved result (no need to recalculate)
        if self.__analysis is None:
            # Expensive computation happens only once
            self.__analysis = self.__perform_analysis()
        return self.__analysis

    def __perform_analysis(self):
        # Complex analysis code
        return sum(self.__data) / len(self.__data)

In [ ]:
# Create analyzer with data
analyzer = DataAnalyzer([10, 20, 30, 40, 50])
print("Analyzer created. No analysis done yet.")

# First time accessing analysis
print("\nFirst access:")
result1 = analyzer.analysis  # Triggers calculation: (10+20+30+40+50)/5 = 30
print(f"Result 1: {result1}")
# Now self.__analysis = 30 (result is saved)

# Second time accessing analysis  
print("\nSecond access:")
result2 = analyzer.analysis  # Just returns saved result: 30
print(f"Result 2: {result2}")
# No calculation needed!

# Third time
print("\nThird access:")
result3 = analyzer.analysis  # Still just returns 30
print(f"Result 3: {result3}")

print(f"\nAll results are the same: {result1 == result2 == result3}")

Analyzer created. No analysis done yet.

First access:
Result 1: 30.0

Second access:
Result 2: 30.0

Third access:
Result 3: 30.0

All results are the same: True


**Property Decorators vs Traditional Methods**
* When should you use property decorators versus traditional getter/setter methods? Here are practical guidelines.

**Use property decorators when**:

* The operation is lightweight and feels like simple data access
* You want your code to read naturally as attribute access
* You're providing read-only access to computed values
* You're adding validation to what looks like a simple attribute

**Use traditional methods when**:

* The operation is expensive or time-consuming
* The method needs multiple parameters
* The method name clearly describes an action (like calculate_total() or send_email())
* You're performing a significant operation beyond just getting or setting a value


In [ ]:
# Example showing when methods are better:
class Database:
    def __init__(self):
        self.__connection = None

    # Good use of method - this is an action
    def connect(self, host, port, username, password):
        self.__connection = self.__create_connection(host, port, username, password)

    # Good use of property - simple status check
    @property
    def is_connected(self):
        return self.__connection is not None

In [ ]:
def check_system_health():
    db = Database()
    
    # Quick status check throughout the day
    if db.is_connected:  # Property - instant check
        return "Database: ✅ Online"
    else:
        return "Database: ❌ Offline"

# This runs every 30 seconds
print(check_system_health())  # Fast property access

Database: ❌ Offline


**Combining Properties with Inheritance**

In [ ]:
class Vehicle:
    def __init__(self, brand):
        self.__brand = brand
        self.__speed = 0

    @property
    def speed(self):
        return self.__speed

    @speed.setter
    def speed(self, value):
        if value >= 0:
            self.__speed = value

class Car(Vehicle):
    @property
    def speed(self):
        return super().speed

    @speed.setter
    def speed(self, value):
        max_speed = 200
        if 0 <= value <= max_speed:
            super(Car, self.__class__).speed.fset(self, value)
        else:
            raise ValueError(f"Speed must be between 0 and {max_speed}")

**Polymorphism**

In [61]:
# building a drawing appllication tthat need to calculate the area of 
# different shapes like circles and rectangles.
class Shape: # parent class or base class : defines the interface,
     #every shape should have an area metho
    def area(self): # method to be overridden
          return 0 # default implementation

# create a specfice shape class for rectangle
class Rectangle(Shape): # child class or derived class
    def __init__(self, width , height): # constructor method
        self.width = width # instance attribute
        self.height = height # instance attribute
    def area(self): # overriding method
        return self.width * self.height # area calculation for rectangle
    
# create a specific shape class for circle
class Circle(Shape): # child class or derived class
    def __init__(self, radius): # constructor method
        self.radius = radius  # instance attribute
    def area (self): #overriding method 
        return 3.14159 *self.radius **2 # area calulation for circle
# cereate a specific shape class for triangle
class Triangle(Shape): # child class or derived class
    def __init__(self, base, height): # constructor method
        self.base = base # instance attribute
        self.height = height # instance attribute
    def area(self): # overriding method
        return 0.5 * self.base * self.height # area calculation for triangle
 # polymorphism becomes powerful. 
 # We can write code that works with any shape without knowing the specific type:  
def print_area(shape): # shape parameter expects a Shape object
    result = shape.area() # We can call this because...
    print(f"Area: {shape.area():.2f}") # calls the area method  

# Understanding `shape.area()` as a Method Call

**Key Point:** `shape.area()` is a **method call** in Python, not just an attribute access.

## Method Call Components:
- **`shape`** = The object (instance)
- **`.`** = The dot operator (connects object to method)  
- **`area`** = The method name
- **`()`** = Parentheses (required to actually CALL the method)

## Why Polymorphism Works:
Since all shape classes inherit from `Shape` and override the `area()` method, we can call `shape.area()` on any shape object, and Python will automatically use the correct implementation based on the object's actual type.

In [67]:
# Let's demonstrate that area() is indeed a method call
# Create different shape objects
rectangle = Rectangle(5, 3)
circle = Circle(4)
triangle = Triangle(6, 8)

# Each of these is a METHOD CALL - notice the parentheses ()
print("Method Calls Demonstration:")
print(f"Rectangle area: {rectangle.area():.2f}")  # Calls Rectangle's area() method
print(f"Circle area: {circle.area():.2f}")        # Calls Circle's area() method  
print(f"Triangle area: {triangle.area():.2f}")    # Calls Triangle's area() method

print("\nPolymorphism in Action:")
# The print_area function can work with ANY shape because they all have area() method
print_area(rectangle)  # Calls rectangle.area() internally
print_area(circle)     # Calls circle.area() internally
print_area(triangle)   # Calls triangle.area() internally

Method Calls Demonstration:
Rectangle area: 15.00
Circle area: 50.27
Triangle area: 24.00

Polymorphism in Action:
Area: 15.00
Area: 50.27
Area: 24.00


## Method vs Attribute - Important Distinction

**Method Call:** `shape.area()` with parentheses `()` 
- **Action:** Executes the area calculation  
- **Result:** Returns the computed area value
- **Purpose:** Performs an operation

**Attribute Access:** `shape.radius` without parentheses
- **Action:** Retrieves stored data
- **Result:** Returns the stored value  
- **Purpose:** Gets a piece of data

## Proof that `area()` is a Method:
```python
# You can check the type of area
print(type(circle.area))    # <class 'method'>
print(type(circle.radius))  # <class 'int'> or <class 'float'>
```

The presence of `()` is what makes it a method **call** rather than just method **reference**.

**Working with Collections of Different Types**
* One of the most practical applications of polymorphism is working with collections that contain different types of objects. 
* You can iterate through a list of mixed objects and call the same method on each, letting polymorphism handle the differences.

In [66]:
# Building on our shapes example with a list of shapes
shapes = [Rectangle(5, 3),
           Circle(7), 
          Triangle(6, 8),
          Rectangle(2, 4),
          Circle(5)]
total_area = 0
for shape in shapes:
    total_area += shape.area() # Polymorphic method call
print(f"\nTotal area of all shapes: {total_area:.2f}")


Total area of all shapes: 279.48


**Key Components for loop:** 
1. **for** keyword - Tells Python "I want to repeat something"

2. **shape** - This is the loop variable (you can name it anything)

Each time through the loop, shape holds one item from the list
First iteration: shape = Rectangle(5, 3)
Second iteration: shape = Circle(7)
Third iteration: shape = Triangle(6, 8), etc.
3. in keyword - Connects the loop variable to the collection

4. shapes - The collection you're iterating through (your list of shape objects)

5. : colon - Marks the end of the for loop header

6. Indented code block - Everything indented under the for loop runs once for each item

Why This Works with Polymorphism:

# Each iteration, 'shape' is a different object type:shape.area()  # But ALL shapes have an area() method!
First loop: shape is Rectangle(5,3) → calls Rectangle.area()
Second loop: shape is Circle(7) → calls Circle.area()
Third loop: shape is Triangle(6,8) → calls Triangle.area()

Alternative Names (all work the same):

* for item in shapes:          # Generic
* for s in shapes:             # Short
* for current_shape in shapes: # Descriptive
* for obj in shapes:           # Object-focused

The magic is that Python doesn't care what type each object is - it just calls area() on whatever object shape currently holds, and polymorphism ensures each object responds with its own implementation!

